Source/santad/Logs/EndpointSecurity/Writers/Spool.mm

```diff
@@ -0,0 +1,102 @@
+/// Copyright 2022 Google LLC
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     https://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+
+#import "Source/santad/Logs/EndpointSecurity/Writers/Spool.h"
+
+#import "Source/common/SNTLogging.h"
+#include "Source/common/santa_proto_include_wrapper.h"
+
+static const char *kTypeGoogleApisComPrefix = "type.googleapis.com/";
+
+namespace santa::santad::logs::endpoint_security::writers {
+
+std::shared_ptr<Spool> Spool::Create(std::string_view base_dir, size_t max_spool_disk_size,
+                                     size_t max_spool_batch_size, uint64_t flush_timeout_ms) {
+  dispatch_queue_t q = dispatch_queue_create("com.google.santa.daemon.file_base_q",
+                                             DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+  dispatch_source_t timer_source = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, q);
+  dispatch_source_set_timer(timer_source, dispatch_time(DISPATCH_TIME_NOW, 0),
+                            NSEC_PER_MSEC * flush_timeout_ms, 0);
+
+  auto spool_writer =
+    std::make_shared<Spool>(q, timer_source, base_dir, max_spool_disk_size, max_spool_batch_size);
+
+  spool_writer->BeginFlushTask();
+
+  return spool_writer;
+}
+
+Spool::Spool(dispatch_queue_t q, dispatch_source_t timer_source, std::string_view base_dir,
+             size_t max_spool_disk_size, size_t max_spool_batch_size)
+    : q_(q),
+      timer_source_(timer_source),
+      spool_writer_(base_dir, max_spool_disk_size),
+      log_batch_writer_(&spool_writer_, max_spool_batch_size) {
+  type_url_ = kTypeGoogleApisComPrefix + ::santa::pb::v1::SantaMessage::descriptor()->full_name();
+}
+
+Spool::~Spool() {
+  // Note: `log_batch_writer_` is automatically flushed when destroyed
+  if (flush_task_started_) {
```

Comment:
Sorta... I have a bug here, but it's not about cancellation.

When a `dispatch_source_t` is created, it has a retain count of 0. When the `dispatch_source_t` is destroyed it will be released. If the object wasn't first `dispatch_resumed` to increment the retain count, then the program will crash.

I did this sorta correctly in the `Metrics` wrapper: https://github.com/google/santa/blob/main/Source/santad/Metrics.mm#L71

However there i should also have done a `dispatch_cancel` prior to `dispatch_resume` to ensure the timer is never fired.

The docs say all this in a roundabout way (IMO...) `dispatch_source_cancel(3)`:
```
If a source is canceled before the first time it is resumed, its event handler will never be called.  (In this case, note that the source must be resumed before it can be released.)
```

So the fix here is to make it look more like the `Metrics` class, but in both cases make sure to `cancel` and then `resume`.